proposals/0060-ct-storage.md

```diff
@@ -0,0 +1,496 @@
+    * Remote Write ingestion would need to be persistent and stateful, which blocks horizontal scalability of receiving.
+    * It limits effectiveness of using CT for PromQL operations like `rate`, `resets` etc.
+    * It makes "rolloup" (write time recording rules that pre-calculate rates) difficult to implement.
+  * Given immutability invariant (e.g. Prometheus), you can't effectively inject CT at a later time (out of order writes are sometimes possible, but expensive, especially for a single sample to be written in the past per series).
+  * It's prone to OOO false positives (we ignore this error for CTs now in Prometheus).
+  * It's producing an artificial sample, which looks like it was scraped.
+* We can't implement delta temporarily effectively.
+
+## Goals
+
+* [MUST] Prometheus can reliably store, query, ingest and export cumulative created timestamp (CT) information (long term plan for [PROM-29](https://github.com/prometheus/proposals/blob/main/proposals/0029-created-timestamp.md#:~:text=For%20those%20reasons%2C%20created%20timestamps%20will%20also%20be%20stored%20as%20metadata%20per%20series%2C%20following%20the%20similar%20logic%20used%20for%20the%20zero%2Dinjection.))
+* [SHOULD] Prometheus can reliably store, query, ingest and export delta start time information. This unblocks [PROM-48 delta proposal](https://github.com/prometheus/proposals/pull/48). Notably adding delta feature later on should ideally not require another complex storage design or implementation.
+* [SHOULD] Overhead of the solution should be minimal--initial overhead target set to maximum of 10% CPU, 10% of memory and 15% of disk space.
+* [SHOULD] Improve complexity/tech-debt of TSDB on the way if possible.
+* [SHOULD] Complexity of consuming CTs should be minimal (e.g. low amount of knowledge needed to use it).
+
+## Non-goals
+
+In this proposal we don't want to:
+
+* Expand details on delta temporality. For this proposal it's enough to assume about delta, what's described in the [Background: Delta Temporality](#background-delta-temporality).
+* Expand or motivate the non-monotonic counter feature.
+
+## How
+
+Native CT support may feel like a big change with a lot of implications. However, we have all the technical pieces to do this, especially if we implement CT with
+(more or less) the same guarantees and details as normal sample timestamp (T), just optional.
+
+General decisions and principles affecting smaller technical parts:
+
+1. We propose to stick to the assumption that CT can change for every sample and don't try to prematurely optimize for the special cumulative best case. Rationales:
+
+* This unlocks the most amount of benefits (e.g. also delta) for the same amount of work, it makes code simpler.
+* We don't know if we need special cumulative best case optimization (yet); also it would be also for some "best" cases. Once we know we can always add those optimizations.
```

Comment:
```suggestion
* We don't know if we need special cumulative best case optimization (yet); also, it would be for some "best" cases. Once we know we can always add those optimizations.
```